File: cpp/graph/maximum_flow/EdmondsKarp.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class EdmondsKarp {
public:
    int search(vector<VertexNode *> &graphs, int start, int end) {
        vector<vector<int>> weights(graphs.size(), vector<int>(graphs.size(), 0));
        for (int i = 0; i < (int) graphs.size(); ++i) {
            EdgeNode *edge = graphs[i]->next;
            while (edge) {
                weights[i][edge->index] = edge->weight;
                edge = edge->next;
            }
        }

        vector<int> parent(0);
        vector<int> maxFlow(graphs.size(), INT_MAX);
        int result = 0;
        while (true) {
            int value = search(graphs, start, end, weights, parent, maxFlow);
            if (!~value) {
                break;
            }

            int index = end;
            while (index != start) {
                weights[parent[index]][index] -= value;
                weights[index][parent[index]] += value;
                index = parent[index];
            }
            result += value;
        }
        return result;
    }

private:
    int search(vector<VertexNode *> &graphs, int start, int end, vector<vector<int>> &weights, vector<int> &parent, vector<int> &maxFlow) {
        parent.assign(graphs.size(), -1);
        queue<int> queue;
        queue.push(start);
        while (!queue.empty()) {
            int index = queue.front();
            queue.pop();
            EdgeNode *edge = graphs[index]->next;
            while (edge) {
                int edgeIndex = edge->index;
                if (weights[index][edgeIndex] && !~parent[edgeIndex]) {
                    parent[edgeIndex] = index;
                    maxFlow[edgeIndex] = min(weights[index][edgeIndex], maxFlow[index]);
                    queue.push(edgeIndex);
                }
                edge = edge->next;
            }
        }
        return ~parent[end] ? maxFlow[end] : -1;
    }
};

File: py/graph/maximum_flow/EdmondsKarp.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class EdmondsKarp:
    def search(self, graphs: List[VertexNode], start: int, end: int) -> int:
        weights = [[0 for _ in range(len(graphs))] for _ in range(len(graphs))]
        for i, value in enumerate(graphs):
            edge = value.next
            while edge:
                weights[i][edge.index] = edge.weight
                edge = edge.next

        parent, maxFlow, result = [], [sys.maxsize for _ in range(len(graphs))], 0
        while True:
            value = self.__search(graphs, start, end, weights, parent, maxFlow)
            if not ~value:
                break

            index = end
            while index != start:
                weights[parent[index]][index] -= value
                weights[index][parent[index]] += value
                index = parent[index]
            result += value
        return result

    def __search(self, graphs: List[VertexNode], start: int, end: int, weights: List[List[int]], parent: List[int], maxFlow: List[int]) -> int:
        parent[ : ], queue = [-1 for _ in range(len(graphs))], [start]
        while queue:
            index = queue.pop(0)
            edge = graphs[index].next
            while edge:
                edgeIndex = edge.index
                if weights[index][edgeIndex] and not ~parent[edgeIndex]:
                    parent[edgeIndex], maxFlow[edgeIndex] = index, min(weights[index][edgeIndex], maxFlow[index])
                    queue.append(edgeIndex)
                edge = edge.next
        return maxFlow[end] if ~parent[end] else -1
